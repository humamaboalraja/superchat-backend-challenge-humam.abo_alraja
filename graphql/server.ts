import { ApolloServer } from 'apollo-server-express';
import express from 'express';
import { createServer } from 'http';
import { createContext } from './context';
import { schema } from './schema';

const { PORT = 5000 } = process.env;

const app = express();

const server = createServer(app);

export const apollo = new ApolloServer({
  schema,
  introspection: process.env.STAGE !== 'production',
  apollo: {},
  context: createContext,
});

export async function main() {
  await apollo.start();
  apollo.applyMiddleware({ app });
  server.listen({ port: PORT }, () => {
    process.stdout.write(
      ` ðŸš€  Server ready at http://localhost:${PORT}${apollo.graphqlPath}`
    );
  });
}
process.env.STAGE != 'test' && main();
