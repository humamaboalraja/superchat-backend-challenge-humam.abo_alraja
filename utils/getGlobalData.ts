import { prisma } from '../singletons/prisma';

export default async function getGlobalData() {
  const messages = await prisma.message.findMany({});
  const contacts = await prisma.contact.findMany({});
  if (!messages || !contacts) throw new Error('Unable to set up global data');
  return { messages, contacts };
}
