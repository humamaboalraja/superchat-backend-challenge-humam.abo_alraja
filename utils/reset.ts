import { prisma } from '../singletons/prisma';

export default async function reset() {
  try {
    const models: string[] = ['Message', 'Contact'];
    for (const model in models) {
      await prisma.$executeRaw(`DELETE FROM "${models[model]}";`);
    }
  } catch (e) {
    throw new Error(`${e}`);
  }
}
