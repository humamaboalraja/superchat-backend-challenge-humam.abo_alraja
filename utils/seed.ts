import cryptoRandomString from 'crypto-random-string';
import { prisma } from '../singletons/prisma';

export default async function seed() {
  try {
    const contact1 = await prisma.contact.create({
      data: {
        name: 'Humam Abo Alraja',
        email: 'humam@company.com',
      },
    });
    const contact2 = await prisma.contact.create({
      data: {
        name: 'John',
        email: 'john@company.com',
      },
    });

    const contact3 = await prisma.contact.create({
      data: {
        name: 'Emma Mockassin',
        email: 'emma@company.com',
      },
    });

    const contact4 = await prisma.contact.create({
      data: {
        name: 'Alex Smith',
        email: 'alex@company.com',
      },
    });

    if (!contact1 || !contact2 || !contact3 || !contact4)
      throw new Error('no contacts');

    const one = await prisma.message.create({
      data: {
        messageContent: 'Hey Emma, How is it going?!',
        sender: {
          connect: {
            id: contact1.id,
          },
        },
        reciever: {
          connect: {
            id: contact3.id,
          },
        },
      },
    });

    const two = await prisma.message.create({
      data: {
        messageContent: "Hi Humam, it's going great! How about you?",
        sender: {
          connect: {
            id: contact3.id,
          },
        },
        reciever: {
          connect: {
            id: contact1.id,
          },
        },
      },
    });

    const three = await prisma.message.create({
      data: {
        messageContent:
          'Hey Alex, do you mind sending me the pitch ASAP? Thanks in advance!',
        sender: {
          connect: {
            id: contact3.id,
          },
        },
        reciever: {
          connect: {
            id: contact4.id,
          },
        },
      },
    });

    const four = await prisma.message.create({
      data: {
        messageContent: 'Hi Emma, Definitely!',
        sender: {
          connect: {
            id: contact4.id,
          },
        },
        reciever: {
          connect: {
            id: contact3.id,
          },
        },
      },
    });
  } catch (e) {
    throw new Error(`${e}`);
  }
}
