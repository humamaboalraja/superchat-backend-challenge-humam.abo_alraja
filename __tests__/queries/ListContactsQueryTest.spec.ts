import getGlobalData from '../../utils/getGlobalData';
import reset from '../../utils/reset';
import seed from '../../utils/seed';
import server from '../server';

beforeAll(async () => {
  await reset();
  await seed();
  global.testData = await getGlobalData();
});

describe('Test Contact queries', () => {
  const CONTACTS_QUERY = `
  query Query {
   listContacts {
     name
     email
   }
 }
`;

  it('should list all contacts ', async () => {
    const res = await server.executeOperation({
      query: CONTACTS_QUERY,
    });
    expect(res).toMatchSnapshot();
  });
});
