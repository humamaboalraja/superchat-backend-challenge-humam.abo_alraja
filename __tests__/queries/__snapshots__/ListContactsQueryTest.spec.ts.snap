// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Contact queries should list all contacts  1`] = `
Object {
  "data": Object {
    "listContacts": Array [
      Object {
        "createdAt": "1637612483916",
        "email": "humam@company.com",
        "id": "ckwb46vkc0000m1hf9823y9qe",
        "name": "Humam Abo Alraja",
        "updatedAt": "1637612483917",
      },
      Object {
        "createdAt": "1637612483932",
        "email": "john@company.com",
        "id": "ckwb46vks0007m1hfuq4e2eaz",
        "name": "John",
        "updatedAt": "1637612483933",
      },
      Object {
        "createdAt": "1637612483942",
        "email": "emma@company.com",
        "id": "ckwb46vl20014m1hfx3d51ars",
        "name": "Emma Mockassin",
        "updatedAt": "1637612483943",
      },
      Object {
        "createdAt": "1637612483951",
        "email": "alex@company.com",
        "id": "ckwb46vlb0021m1hf3khm4m99",
        "name": "Alex Smith",
        "updatedAt": "1637612483952",
      },
    ],
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;
